play-json/jvm/src/main/scala/play/api/libs/json/EnvReads.scala

```diff
@@ -415,6 +419,82 @@ trait EnvReads {
     case _ => JsError(JsonValidationError("error.expected.jsstring"))
   }
 
+  private def javaDurationNumberReads(unit: TemporalUnit, esuffix: String) =
+    Reads[JDuration] {
+      case JsNumber(n) if !n.ulp.isWhole =>
+        JsError(s"error.invalid.$esuffix")
+
+      case JsNumber(n) => JsSuccess(JDuration.of(n.toLong, unit))
+      case _ => JsError(s"error.expected.$esuffix")
+    }
+
+  /**
+   * Deserializer of Java Duration from an integer (long) number,
+   * using the specified temporal unit.
+   */
+  def javaDurationNumberReads(unit: TemporalUnit): Reads[JDuration] =
+    javaDurationNumberReads(unit, "duration")
+
+  /** Deserializer of Java Duration from a number of milliseconds. */
+  val javaDurationMillisReads: Reads[JDuration] =
+    javaDurationNumberReads(ChronoUnit.MILLIS)
+
+  /**
+   * Deserializer of Java Duration, from either a time-based amount of time
+   * (string representation such as '34.5 seconds'),
+   * or from a number of milliseconds (see [[javaDurationMillisReads]]).
+   */
+  implicit val DefaultJavaDurationReads: Reads[JDuration] = Reads[JDuration] {
+    case JsString(repr) => try {
+      JsSuccess(JDuration.parse(repr))
+    } catch {
+      case _: Exception => JsError("error.invalid.duration")
+    }
+
+    case js => javaDurationMillisReads.reads(js)
+  }
+
+  // Should only be called with a unit supported by Period
+  private def javaPeriodNumberReads(f: Int => Period): Reads[Period] =
+    Reads[Period] {
+      case JsNumber(n) if !n.isValidInt =>
+        JsError(s"error.invalid.period")
```

Comment:
Another note about message keys: I would expect the error to be different if I'm reading a number versus if I'm reading as a period string. I'd assume `error.invalid.period` would correspond to a message like "Invalid period", which is not enough information for me to understand why this is failing.